pdns/dnsdistdist/dnsdist-lbpolicies.cc

```diff
@@ -31,58 +31,68 @@ static constexpr size_t s_staticArrayCutOff = 16;
 template <typename T> using DynamicIndexArray = std::vector<std::pair<T, size_t>>;
 template <typename T> using StaticIndexArray = std::array<std::pair<T, size_t>, s_staticArrayCutOff>;
 
-template <class T> static std::shared_ptr<DownstreamState> getLeastOutstanding(const ServerPolicy::NumberedServerVector& servers, T& poss)
+static std::optional<ServerPolicy::SelectedServerPosition> getLeastOutstanding(const ServerPolicy::NumberedServerVector& servers)
 {
-  /* so you might wonder, why do we go through this trouble? The data on which we sort could change during the sort,
-     which would suck royally and could even lead to crashes. So first we snapshot on what we sort, and then we sort */
-  size_t usableServers = 0;
-  for (const auto& d : servers) {
-    if (d.second->isUp()) {
-      poss.at(usableServers) = std::pair(std::tuple(d.second->outstanding.load(), d.second->d_config.order, d.second->getRelevantLatencyUsec()), d.first);
-      usableServers++;
+  std::optional<ServerPolicy::SelectedServerPosition> best;
+  uint64_t leastOutstandingSeen = std::numeric_limits<uint64_t>::max();
+  int lowestOrderSeen = std::numeric_limits<int>::max();
+  double lowestLatencySeen = std::numeric_limits<double>::max();
+
+  for (const auto& server : servers) {
+    if (!server.second->isUp()) {
+      continue;
     }
-  }
 
-  if (usableServers == 0) {
-    return shared_ptr<DownstreamState>();
+    auto outstanding = server.second->outstanding.load();
+    auto order = server.second->d_config.order;
+    if (outstanding > leastOutstandingSeen) {
+      continue;
+    }
+    if (outstanding < leastOutstandingSeen) {
+      best = server.first;
+      leastOutstandingSeen = outstanding;
+      lowestOrderSeen = order;
+      lowestLatencySeen = server.second->getRelevantLatencyUsec();
+      continue;
+    }
+    if (order > lowestOrderSeen) {
+      continue;
+    }
+    auto latency = server.second->getRelevantLatencyUsec();
+    if (latency < lowestLatencySeen) {
+      best = server.first;
+      leastOutstandingSeen = outstanding;
+      lowestOrderSeen = order;
+      lowestLatencySeen = server.second->getRelevantLatencyUsec();
+    }
   }
 
-  std::nth_element(poss.begin(), poss.begin(), poss.begin() + usableServers, [](const typename T::value_type& a, const typename T::value_type& b) { return a.first < b.first; });
-  // minus 1 because the NumberedServerVector starts at 1 for Lua
-  return servers.at(poss.begin()->second - 1).second;
+  return best;
 }
 
 // get server with least outstanding queries, and within those, with the lowest order, and within those: the fastest
-shared_ptr<DownstreamState> leastOutstanding(const ServerPolicy::NumberedServerVector& servers, const DNSQuestion* dq)
+std::optional<ServerPolicy::SelectedServerPosition> leastOutstanding(const ServerPolicy::NumberedServerVector& servers, const DNSQuestion* dnsQuestion)
 {
-  (void)dq;
-  using LeastOutstandingType = std::tuple<int,int,double>;
+  (void)dnsQuestion;
```

Comment:
if it is unused, perhaps remove this from the function parameters?

Reply:
It is used by some load-balancing policies, and they all need to have the same signature.

Reply:
I prefer `[[maybe_unused]]` for these cases.